eva/readers/opencv_reader.py

```diff
@@ -44,12 +53,26 @@ def _read(self) -> Iterator[Dict]:
         else:
             range_list = [(0, num_frames - 1)]
         logger.debug("Reading frames")
-
-        for (begin, end) in range_list:
-            video.set(cv2.CAP_PROP_POS_FRAMES, begin)
-            _, frame = video.read()
-            frame_id = begin
-            while frame is not None and frame_id <= end:
-                yield {"id": frame_id, "data": frame}
+        if self._sampling_rate == 1:
+            for (begin, end) in range_list:
+                video.set(cv2.CAP_PROP_POS_FRAMES, begin)
                 _, frame = video.read()
-                frame_id += 1
+                frame_id = begin
+                while frame is not None and frame_id <= end:
+                    yield {"id": frame_id, "data": frame}
+                    _, frame = video.read()
+                    frame_id += 1
+        else:
+            for begin, end in range_list:
+
+                # align begin with sampling rate
+                if begin % self._sampling_rate:
+                    begin += self._sampling_rate - (begin % self._sampling_rate)
```

Comment:
If sampling has higher precedence, should it happen before predicate? Is that due to the fact that we only push down range id predicates, so the order does not matter after we align the `begin`?

Reply:
Yes. And I also removed the PushFilterBelowSample Rule to enforce it